tsdb/ooo_head_read.go

```diff
@@ -174,6 +186,60 @@ type multiMeta struct {
 	metas          []chunks.Meta
 }
 
+// Wrapper to return a chunk where the iterator clears
+// counter resets.
+type hintClearedChunk struct {
+	chunkenc.Chunk
+}
+
+func (h hintClearedChunk) Iterator(it chunkenc.Iterator) chunkenc.Iterator {
+	if itt, ok := it.(hintClearedIterator); ok {
+		itt.it = h.Chunk.Iterator(itt.it)
+		return itt
+	}
+	return &hintClearedIterator{it: h.Chunk.Iterator(nil)}
+}
+
+type hintClearedIterator struct {
+	it chunkenc.Iterator
+}
+
+func (h hintClearedIterator) Next() chunkenc.ValueType {
+	return h.it.Next()
+}
+
+func (h hintClearedIterator) Seek(t int64) chunkenc.ValueType {
+	return h.it.Seek(t)
+}
+
+func (h hintClearedIterator) AtT() int64 {
+	return h.it.AtT()
+}
+
+func (h hintClearedIterator) At() (int64, float64) {
+	return h.it.At()
+}
+
+func (h hintClearedIterator) AtHistogram(to *histogram.Histogram) (int64, *histogram.Histogram) {
+	t, H := h.it.AtHistogram(to)
+	if H != nil && H.CounterResetHint == histogram.CounterReset {
+		H.CounterResetHint = histogram.UnknownCounterReset
+	}
```

Comment:
I'm wondering whether it would be safer to always set the first sample as `UnknownCounterReset` rather than checking for `CounterReset` and overriding that. Right now OOO chunks can only have a counter reset header of `UnknownCounterReset` or `CounterReset` so this does work. My concern is that if we ever change the logic and start getting OOO chunks with `NotCounterReset` headers, we wouldn't realise this code would need changing too. That would be more complicated to implement though, since we don't know the position of the returned sample in the underlying chunk.

Reply:
Resolved as headers are now set in ToEncodedChunks()